GPII-2933 Common Event Bindings

sgithens opened the pull request:
@amb26 @the-t-in-rtf I'm going to look at adding a few unit tests, but if you wanted to take an initial look at this and give me any feedback, that'd be helpful

sgithens committed:
GPII-2933 Common Event Bindings

the-t-in-rtf commented:
Great idea, and very useful, as we do this kind of grunt work all the time in downstream grades.  Although I could look at stuff in isolation, it'd be helpful to see tests so that I can see examples of what you're passing and what the intended effect is.  I'm happy with these being synthetic testem-friendly tests, simulating keyboard input crudely rather than using gpii-webdriver.

Also, linting!  Dude! (I was wondering how the heck you managed to break the tests without adding tests for your work.  That's how.)

sgithens committed:
Merge remote-tracking branch 'GPII/master' into GPII-2933

sgithens committed:
GPII-2933 Adding unit tests for common event bindings

sgithens commented:
Hi @the-t-in-rtf .  I added a few unit tests for this, and everything should lint now.  :)

the-t-in-rtf commented on tests/static/js/jqunit/jqUnit-binder-commonEventBindings.js:
```diff
@@ -0,0 +1,92 @@
+                    sequence: [
+                        {
+                            func: "gpii.tests.binder.clickSelector",
+                            args: [".parapgraph-click"]
```
"parapgraph" needs to be updated.

sgithens replied:
thanks :p

the-t-in-rtf commented on src/js/common-event-bindings.js:
```diff
@@ -0,0 +1,95 @@
+    // The methods below might be generic enough to go straight to infusion
+    //
+
+    fluid.expandCompoundArg = function (that, arg, name) {
```
I see your note explaining why these are in the fluid namespace.  However, I assume wherever they're heading they'll need JSDocs.

sgithens replied:
Add JSDocs

the-t-in-rtf commented on tests/static/js/jqunit/jqUnit-binder-commonEventBindings.js:
```diff
@@ -0,0 +1,92 @@
+            paragraphClick: ".parapgraph-click"
+        },
+        markupEventBindings: {
+            inputButton: {
```
You really need to add examples that include multiple bindings for a single element.  For example, I often want to add `click` and `keydown` bindings for elements so that keyboard and mouse navigation are both supported.

sgithens replied:
Interesting, it didn't actually have support for this, but it makes total sense. Added impl and tests to make this possible (without making duplicate selectors)

the-t-in-rtf replied:
Great, I had my own one-offs for this and it'll be nice to have an alternative.

the-t-in-rtf commented on src/js/common-event-bindings.js:
```diff
@@ -0,0 +1,95 @@
+        return expanded;
+    };
+
+    fluid.processjQueryDecorator = function (dec, node, that, name) {
```
One thing  @amb26 and I have been doing lately is using JSDocs [`@typedef` syntax](https://jsdoc.app/tags-typedef.html) to document deeper structures like `dec` here.

sgithens replied:
Added `@typedef` for these

the-t-in-rtf replied:
Much better, I added a small comment on the return value, otherwise, looks good.

the-t-in-rtf commented:
Hi, @sgithens, I'm trying to make sense of [the associated ticket](https://issues.gpii.net/browse/GPII-2933), you kind of describe in the title the change you're making, but I couldn't get more info as most of the links to example there are broken.  I think I get it, but for posterity we need a clear summary.

Also, the new functionality needs to be described somewhere, perhaps in a new section towards the end of the README.

sgithens committed:
GPII-2933 Fixing typo's and starting tests for keydown events.

sgithens committed:
GPII-2933 Adding support for multiple events in a single declaration

sgithens committed:
GPII-2933 JSDocs, README doco, minor grammar fixes.

sgithens commented:
@the-t-in-rtf Updated the README and ticket. I believe this should be ready for another review now.

the-t-in-rtf commented on README.md:
```diff
@@ -177,6 +177,47 @@ If your component generates or regenerates markup, you will need to call `gpii.b
+
+    markupEventBindings: {
+        productListLinks: {
+            method: ["click", "keydown"]
```
Very nice.

the-t-in-rtf commented on src/js/common-event-bindings.js:
```diff
@@ -0,0 +1,162 @@
+     * want to listen for multiple events in the same binding this can be an array of event
+     * types such as `["click", "keypress"]`.
+     * @property {Array} args - A list of arguments to be passed to the event. This supports the
+     * usual range of fluid IOC syntax. Typically, this will be an invoker on the component to be
```
Looks pretty good, but maybe `Fluid IoC`?

the-t-in-rtf commented on src/js/common-event-bindings.js:
```diff
@@ -0,0 +1,162 @@
+     *
+     * @param {Object} that - The component itself.
+     * @param {String|Object} arg - A single argument being passed to the event
+     * infoker.
```
I'm going to start saying `infoker` from now on, but you should still change it.

the-t-in-rtf replied:
Also, I think it can stay on one line?

the-t-in-rtf commented on src/js/common-event-bindings.js:
```diff
@@ -0,0 +1,162 @@
+     * @param {Object} that - The component itself.
+     * @param {String} name - Name that will be used for the invoker created to handle
+     * this event.
+     * @return {Array} List of event functions created for this binding decorator.
```
Maybe add a sub-type here, i.e. is it an array of Functions or an array of String function names?

the-t-in-rtf commented on src/js/common-event-bindings.js:
```diff
@@ -0,0 +1,162 @@
+     * Function to process the markup binding decorators and create the events described
+     * by them. The markup needs to be rendered and settled before this can be called.
+     *
+     * @param {Object} that - The component itself.
```
I am the guiltiest party here, I do this all the time, but in later work, @amb26 and I have been adding the grade to these for purposes of clarity.

sgithens replied:
Make these more specific using the component grade for the type.

the-t-in-rtf commented:
I left a few comments, but am happy with this.  @amb26, can you look through this as well?  I know you and @sgithens have already discussed this, it'd be good to confirm this roughly matches what you had in mind.

amb26 commented on src/js/common-event-bindings.js:
```diff
@@ -0,0 +1,162 @@
+                // TODO: perhaps a a courtesy we could expose {node} or even {this}
+                expanded = fluid.makeInvoker(that, invokerec, name);
+            } else {
+                expanded = fluid.expandOptions(arg, that);
```
As per recent updates in the core this should really be fluid.expandImmediate (but this is also backwards compatible)

sgithens replied:
updated

amb26 commented on tests/static/js/jqunit/jqUnit-binder-commonEventBindings.js:
```diff
@@ -0,0 +1,128 @@
+                method: "click",
+                args: ["{that}.handleParagraphClick"]
+            },
+            multipleEventInputButton: {
```
Whilst we should support this form, we also need to support multiple bindings which don't share a handler, so please add a test which shows support for an array or hash as the RHS here

sgithens replied:
added support for an array of bindings here

amb26 commented on tests/static/js/jqunit/jqUnit-binder-commonEventBindings.js:
```diff
@@ -0,0 +1,128 @@
+            paragraphClick: ".paragraph-click",
+            multipleEventInputButton: "#input-multiple-events-button"
+        },
+        markupEventBindings: {
```
Also supply a test which shows that these handlers can be overridden cleanly

sgithens replied:
Added a test where 1) Some are left as is, 2) some are overridden, and 3) some new one is introduced.

amb26 commented on tests/static/js/jqunit/jqUnit-binder-commonEventBindings.js:
```diff
@@ -0,0 +1,128 @@
+        markupEventBindings: {
+            inputButton: {
+                method: "click",
+                args: ["{that}.handleInputClick"]
```
Supply further tests to validate the primitive kinds of "compact" IoC syntax we support which allows for args

sgithens replied:
Thanks for this tip, I didn't know this was possibly actually, and will simplify my usage in the capture tool, PPT, and elsewhere

sgithens committed:
GPII-2933 Updating jsdocs

sgithens committed:
GPII-2933 Cleanup from PR
- Adding ability to pass an array of handlers to eventMarkupBindings
- Adding unit tests for overriding bindings in sub grades
- Adding unit tests for compact IoC invocation
- Grammar and jsdoc fixes

sgithens commented:
@amb26 @the-t-in-rtf I believe this should be ready for another round of review now

the-t-in-rtf commented:
The build failed because the VM couldn't be created:

```
+ vagrant up --provider virtualbox
The provider 'virtualbox' that was requested to back the machine
'windows10' is reporting that it isn't usable on this system. The
reason is shown below:

Vagrant has detected that you have a version of VirtualBox installed
that is not supported by this version of Vagrant. Please install one of
the supported versions listed below to use Vagrant:

4.0, 4.1, 4.2, 4.3, 5.0, 5.1, 5.2
```

OK to test again.

the-t-in-rtf commented:
Vagrant timeout this go around.  Ok to test again, if it fails a third time we should rope in Alfredo to assist.

the-t-in-rtf commented:
```
==> windows10: ok 13 Firefox 73.0 - [734 ms] - Binder Tests: /tests/static/tests-binder-commonEventBindingsOverride.html
Build was aborted
Aborted by unknown
/opt/vagrant/embedded/gems/2.2.7/gems/winrm-2.3.4/lib/winrm/http/response_handler.rb:59:in `raise_if_auth_error': WinRM::WinRMAuthorizationError (WinRM::WinRMAuthorizationError)
	from /opt/vagrant/embedded/gems/2.2.7/gems/winrm-2.3.4/lib/winrm/http/response_handler.rb:51:in `raise_if_error'
```

Looks like the build was killed after timing out, I suspect issues with testem.  I am testing locally at the moment and will advise shortly on how to proceed.

the-t-in-rtf commented:
OK, there are two problems:

1. You need to merge with upstream master to pick up the fix for the Firefox hang issues described in [GPII-4145](https://issues.gpii.net/browse/GPII-4145).
2. Since upstream master now uses a potentia-ii version of Infusion, you'll need to update your dependencies in the new test fixtures to use the same infusion-all rollup that the rest of the tests use.  Otherwise you're missing the now-required promises API.

With those two fixes your tests pass in Vagrant locally.

sgithens committed:
Merge remote-tracking branch 'GPII/master' into GPII-2933

sgithens committed:
GPII-2933 Updated to infusion dist to match other tests.

sgithens commented:
@the-t-in-rtf Thanks for the sluething, try these updates now.

sgithens commented:
@the-t-in-rtf Yes!  It passed

sgithens committed:
GPII-2933 Updating that types in jsdocs

sgithens commented:
Looks like a timeout...

ok to test

sgithens commented:
@the-t-in-rtf Ok, this is passed, and I think ready for another review now.

the-t-in-rtf commented on src/js/common-event-bindings.js:
```diff
@@ -0,0 +1,165 @@
+     */
+    fluid.defaults("gpii.binder.bindMarkupEvents", {
+        mergePolicy: {
+            decorators: "noexpand"
```
Why is this "decorators"?

the-t-in-rtf replied:
In other words, is there a pattern we're emulating?  Just curious.

the-t-in-rtf commented on src/js/common-event-bindings.js:
```diff
@@ -0,0 +1,165 @@
+            onMarkupRendered: null
+        },
+        listeners: {
+            onMarkupRendered: "{that}.events.onDomBind.fire({that}, {that}.container)",
```
These should all be namespaced.  I'd also suggest using the "long form" to this shorter notation.  That way someone can at least opt out of part of the contract in a derived grade by changing the "funcName" to `fluid.identity` or the like.

the-t-in-rtf commented on src/js/common-event-bindings.js:
```diff
@@ -0,0 +1,165 @@
+            if (node.length > 0) {
+                var name = "Decorator for DOM node with selector " + key + " for component " + fluid.dumpThat(that);
+                // val can be an array to support multiple event handlers
+                var handlerDecs = fluid.isArrayable(val) ? val : [val];
```
Why not just use `fluid.makeArray` here?

the-t-in-rtf commented on tests/static/js/jqunit/jqUnit-binder-commonEventBindings.js:
```diff
@@ -0,0 +1,158 @@
+            }
+        },
+        listeners: {
+            "onCreate.markupEventBindings": "{that}.events.onMarkupRendered.fire()"
```
This is the kind of thing I'd also write out, what if someone needs to do other work before firing the event?  I guess they can just rewrite the whole string.

the-t-in-rtf commented on tests/static/js/jqunit/jqUnit-binder-commonEventBindings.js:
```diff
@@ -0,0 +1,158 @@
+        }
+    });
+
+    gpii.tests.binder.commonEventBindings.environment();
```
I guess it matters less on the browser side (I think?), but for consistency with the rest of the package, this should use `fluid.test.runTests` instead of calling the environment directly.

the-t-in-rtf commented on tests/static/js/jqunit/jqUnit-binder-commonEventBindingsOverride.js:
```diff
@@ -0,0 +1,177 @@
+        }
+    });
+
+    gpii.tests.binder.commonEventBindingsOverride.environment();
```
We should either consistently use `fluid.test.runTests` or consistently not use it.

amb26 replied:
+1 since without runTests there is no teardown

the-t-in-rtf commented on tests/static/tests-binder-commonEventBindingsOverride.html:
```diff
@@ -0,0 +1,53 @@
+        <script type="text/javascript" src="../../src/js/common-event-bindings.js"></script>
+    </head>
+    <body>
+        <h1 id="qunit-header">"Common Event Bindings" Component Tests</h1>
```
Please update the page and header title to distinguish this from the other tests.

the-t-in-rtf commented on tests/static/tests-binder-commonEventBindingsOverride.html:
```diff
@@ -0,0 +1,53 @@
+        <!-- Test HTML, we have to hide this ourselves.  QUnit's styles put it offscreen, which doesn't work for us. -->
+        <div class="viewport-common-event-bindings" style="display:none">
+            <form>
+                <h3>Common Event Bindings...</h3>
```
I know it's mostly harmless, but for any human trying to troubleshoot the test, this should also be different than the other comment event bindings tests.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/fluid-project/fluid-binder?pullRequest=13) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/fluid-project/fluid-binder?pullRequest=13) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/fluid-project/fluid-binder?pullRequest=13) it.</sub>
